### PR TITLE
Introduce OpenTelemetry observability service

### PR DESCRIPTION
**Pull request by joaopgrassi**
This PR introduces a new observability service that leverages OpenTelemetry instrumentation around cloudevents clients:

- Added new observability service
- Tests and samples
- Documentation

### CODE REVIEW

File: observability/opentelemetry/v2/client/cloudevents_carrier.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+import (
+	"context"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/extensions"
+)
+
+// CloudEventCarrier adapts the distributed trace extension to satisfy the TextMapCarrier interface.
+// https://github.com/open-telemetry/opentelemetry-go/blob/main/propagation/propagation.go#L23
```
Comment (arminru) [inline observability/opentelemetry/v2/client/cloudevents_carrier.go]:
```suggestion
https://github.com/open-telemetry/opentelemetry-go/blob/v1.0.0-RC3/propagation/propagation.go#L23
```
Otherwise the line reference might become wrong (you can either select a branch from the dropdown on top or press `y` to have the branch replaced by the commit hash but I find the branch reference more expressive).

File: observability/opentelemetry/v2/client/cloudevents_carrier.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+package client
+
+import (
+	"context"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/extensions"
+)
+
+// CloudEventCarrier adapts the distributed trace extension to satisfy the TextMapCarrier interface.
```
Comment (arminru) [inline observability/opentelemetry/v2/client/cloudevents_carrier.go]:
```suggestion
// CloudEventCarrier wraps the distributed trace extension to satisfy the TextMapCarrier interface.
```
maybe?

File: observability/opentelemetry/v2/client/cloudevents_trace_context.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+package client
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// CloudEventTraceContext is a wrapper around the OpenTelemetry TraceContext
```
Comment (arminru) [inline observability/opentelemetry/v2/client/cloudevents_trace_context.go]: Why is this necessary?

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+// NewOTelObservabilityService returns a OpenTelemetry enabled observability service
+func NewOTelObservabilityService(opts ...OTelObservabilityServiceOption) *OTelObservabilityService {
+	tracerProvider := otel.GetTracerProvider()
+
+	os := &OTelObservabilityService{
+		tracer: tracerProvider.Tracer(
+			instrumentationName,
+			trace.WithInstrumentationVersion("1.0.0"), // TODO: Can we have the package version here?
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go]: This should be the package version, yes. Otherwise you could just leave it out.

File: observability/opentelemetry/v2/client/otel_options.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"go.opentelemetry.io/otel/attribute"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"github.com/cloudevents/sdk-go/v2/observability"
+)
+
+const (
+	// TODO: Should this be the name of this module or the whole package?
+	instrumentationName = "github.com/cloudevents/sdk-go/observability/opentelemetry/v2"
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_options.go:17]: I _think_ the whole package would be more applicable unless you'd add separate instrumentation right into the SDK itself.

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+)
+
+// OTelObservabilityService implements the ObservabilityService interface from cloudevents
+type OTelObservabilityService struct {
+	tracer               trace.Tracer
+	spanAttributesGetter func(cloudevents.Event) []attribute.KeyValue
+	spanNameFormatter    func(cloudevents.Event) string
+}
+
+// NewOTelObservabilityService returns a OpenTelemetry enabled observability service
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go]:
```suggestion
// NewOTelObservabilityService returns an OpenTelemetry-enabled observability service
```

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+	return []func(context.Context, binding.Message) context.Context{tracePropagatorContextDecorator}
+}
+
+// RecordReceivedMalformedEvent records the error from a malformed event in the span.
+func (os OTelObservabilityService) RecordReceivedMalformedEvent(ctx context.Context, err error) {
+	spanName := observability.ClientSpanName + ".malformed receive"
+	_, span := os.tracer.Start(
+		ctx, spanName,
+		trace.WithSpanKind(trace.SpanKindConsumer),
+		trace.WithAttributes(attribute.String(string(semconv.CodeFunctionKey), "RecordReceivedMalformedEvent")))
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go]: Is there no way to reference it as a symbol that would survive renamings?

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+// InboundContextDecorators returns a decorator function that allows enriching the context with the incoming parent trace.
+// This method gets invoked automatically by passing the option 'WithObservabilityService' when creating the cloudevents HTTP client.
+func (os OTelObservabilityService) InboundContextDecorators() []func(context.Context, binding.Message) context.Context {
+	return []func(context.Context, binding.Message) context.Context{tracePropagatorContextDecorator}
+}
+
+// RecordReceivedMalformedEvent records the error from a malformed event in the span.
+func (os OTelObservabilityService) RecordReceivedMalformedEvent(ctx context.Context, err error) {
+	spanName := observability.ClientSpanName + ".malformed receive"
+	_, span := os.tracer.Start(
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go]: Looks like if this would be in the `os` package, maybe use `obss` or something like that instead for clarity.

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+		trace.WithAttributes(attribute.String(string(semconv.CodeFunctionKey), "RecordReceivedMalformedEvent")))
+
+	recordSpanError(span, err)
+	span.End()
+}
+
+// RecordCallingInvoker starts a new span before calling the invoker upon a received event.
+// In case the operation fails, the error is recorded and the span is marked as failed.
+func (os OTelObservabilityService) RecordCallingInvoker(ctx context.Context, event *cloudevents.Event) (context.Context, func(errOrResult error)) {
+	spanName := os.getSpanName(event, "receive")
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go]:
```suggestion
	spanName := os.getSpanName(event, "process")
```
This is not tracking busy waiting for a message but the handler function processing it, right?
See https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md.

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+	}
+}
+
+func GetDefaultSpanAttributes(e *cloudevents.Event, method string) []attribute.KeyValue {
+	attr := []attribute.KeyValue{
+		attribute.String(string(semconv.CodeFunctionKey), method),
+		attribute.String(observability.SpecversionAttr, e.SpecVersion()),
+		attribute.String(observability.IdAttr, e.ID()),
+		attribute.String(observability.TypeAttr, e.Type()),
+		attribute.String(observability.SourceAttr, e.Source()),
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go:138]: These should be added to the OTel semconv as well:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#attributes-specific-to-certain-messaging-systems

File: observability/opentelemetry/v2/client/otel_observability_service.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+		span.SetAttributes(os.spanAttributesGetter(event)...)
+	}
+
+	return ctx, func(errOrResult error, event *cloudevents.Event) {
+		recordSpanError(span, errOrResult)
+		span.End()
+	}
+}
+
+func GetDefaultSpanAttributes(e *cloudevents.Event, method string) []attribute.KeyValue {
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_observability_service.go:132]:
```suggestion
func getDefaultSpanAttributes(e *cloudevents.Event, method string) []attribute.KeyValue {
```

File: observability/opentelemetry/v2/client/otel_options.go [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+type OTelObservabilityServiceOption func(*OTelObservabilityService)
+
+func WithSpanAttributesGetter(attrGetter func(cloudevents.Event) []attribute.KeyValue) OTelObservabilityServiceOption {
```
Comment (arminru) [inline observability/opentelemetry/v2/client/otel_options.go:23]: These could use some docs, e.g., here saying that those are added to the built-in attributes and don't replace them.

File: observability/opentelemetry/v2/http/http.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"net/http"
+
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+
+	cehttp "github.com/cloudevents/sdk-go/v2/protocol/http"
+)
+
+// NewObservedHTTP creates an HTTP protocol with trace propagating middleware.
```
Comment (arminru) [inline observability/opentelemetry/v2/http/http.go]:
```suggestion
// NewObservedHTTP creates an HTTP protocol with OTel trace propagating middleware.
```

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 9 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: Yeah this is invoked when the busy-waiting process receives the event, right before calling the handler function. 

With the client you get from cloudevents, you can listen (blocking) like this: https://github.com/cloudevents/sdk-go#receive-your-first-cloudevent. 

Then this method on the obs service is called from the invoker: https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/client/invoker.go#L83.

I think both operation names are "appropriate" so I'm not sure which one is more semantic.

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 10 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go:138]: What is the process for that? Just open a PR or should it be discussed first on a issue?

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 11 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go:132]: I left it exported because I use it on the tests to assert the default + additional attributes are as expected. I couldn't make it work to make it exported *only* for the tests, and I didn't want to copy/paste it to the test as it might change in the future.

File: observability/opentelemetry/v2/client/cloudevents_carrier.go (hunk 1 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/cloudevents_carrier.go]: Good point. I changed it here in other places as well.

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 10 above)
Comment (arminru) [reply observability/opentelemetry/v2/client/otel_observability_service.go:138]: If this was brand new it should probably be discussed first on an issue but since CloudEvents themselves already have the attributes defined on their end I think you can just create a PR towards otel-spec and reference `keys.go` from here (or the CloudEvents spec if it's there as well) as source.

File: observability/opentelemetry/v2/client/otel_options.go (hunk 12 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_options.go:23]: Done

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 8 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: Good point. I used simply `o` as the current opencensus on.

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 7 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: Found a way.. not so great as in other languages but now the method name is found at runtime.

File: observability/opentelemetry/v2/client/otel_options.go (hunk 5 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_options.go:17]: That seems reasonable. I will stick with it them :)

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 4 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: Since I'm not sure if we can get it at runtime (my search didn't yield any good results) I removed it for now.

Comment (arminru) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: It would be nice to have though. Maybe they have it stored in a metadata file or constant somewhere. You could leave it commented out but with the TODO so they can tell you.

File: observability/opentelemetry/v2/client/cloudevents_trace_context.go (hunk 3 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/cloudevents_trace_context.go]: Yeah after our discussions and thinking more about I decided to remove it. Just adds more complexity which is not needed. Now using the default from OpenTelemetry.

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 4 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: Right. I left the line commented out, hopefully they can point me on how to get it.

File: observability/opentelemetry/v2/client/otel_observability_service.go (hunk 9 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/otel_observability_service.go]: I changed to `process`. I agree it makes more sense.

File: observability/opentelemetry/v2/README.md [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+c, _ := otelObs.NewClientHTTP([]cehttp.Option{}, []client.Option{})
+```
+
+This will produce spans for all outgoing and incoming requests. By default, the spans will have the attributes as defined in [keys.go](https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/observability/keys.go). For more advanced configuration, see the next section.
+
+## Advanced configuration
+
+### HTTP auto-instrumentation
+
+In order to generate spans when sending and receiving events, it's necessary to configure the HTTP client from cloudevents with OpenTelemetry instrumentation. The client has two "possibilities" for instrumentation:
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
In order to generate spans when sending and receiving events, it's necessary to configure the HTTP client from cloudevents with OpenTelemetry instrumentation. The client has two potentially interesting points for instrumentation:
```
maybe?

File: observability/opentelemetry/v2/README.md [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+	}),
+)
+```
+
+The `otelhttp.NewTransport` will ensure that spans are generated for each outgoing request, and that the `traceparent` header is properly propagated. The `otelhttp.NewHandler` will take care of incoming requests, reading the `traceparent` header and continuing the trace with a new span.
+
+This already gives some observability "out-of-the-box", but the spans generated only contain common HTTP headers as defined in the [HTTP semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/trace/semantic_conventions/http.md). Also if you are using another protocol, then spans will not be automatically generated, unless there is an auto-instrumentation library for that.
+
+Because of this, cloudevents offers the `ObservabilityService` interface which is used to generate spans, independently of the chosen protocol. See next how to configure the cloudevents client to use it.
+
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md:48]:
```suggestion
```

File: observability/opentelemetry/v2/README.md [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+  <tbody>
+    <tr>
+      <td style="font-weight:bold">Span name</td>
+      <td><code>cloudevents.client.[event type] [operation name]</code> where <code>[operation name]</code> follows the OpenTelemetry <a href="https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/trace/semantic_conventions/messaging.md#operation-names">semantic conventions for messaging systems</a></td>
+    </tr>
+    <tr>
+      <td style="font-weight:bold">Span attributes</td>
+      <td>The attributes as defined in <a href="https://github.com/cloudevents/sdk-go/blob/release-2.5/v2/observability/keys.go">keys.go</a></td>
+    </tr>
+    </tbody>
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
  </tbody>
```

File: observability/opentelemetry/v2/README.md [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+```
+
+>NOTE: The `nameFormatter` and `attributesGetter` functions will be called on each span creation. **Avoid** doing any heavy processing in them.
+
+
+## Extra types
+
+This package also contains extra types and helper functions that are useful in case you need to access/set the `tracecontext` in a more "low-level" way.
+
+They allow to inject(write) and extract(read) `tracecontext` from the event. This is particularly useful when dealing with code that has no notion of a "request" nor a context. For example, long-running background processes pooling from a queue.
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
They allow to inject(write) and extract(read) `tracecontext` from the event. This is particularly useful when dealing with code that has no notion of a "request" nor a context. For example, long-running background processes polling from a queue. 
```
I might be wrong about which verb you wanted to use here, so please ignore if this is not what you meant :sweat_smile:

File: observability/opentelemetry/v2/README.md [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+// create the obs service with custom span names and attributes
+os := otelObs.NewOTelObservabilityService(
+	otelObs.WithSpanNameFormatter(nameFormatter),
+	otelObs.WithSpanAttributesGetter(attributesGetter),
+)
+
+c, err := cloudevents.NewClient(p, client.WithObservabilityService(os))
+```
+
+>NOTE: The `nameFormatter` and `attributesGetter` functions will be called on each span creation. **Avoid** doing any heavy processing in them.
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
>Note: The `nameFormatter` and `attributesGetter` functions will be called on each span creation. **Avoid** doing any heavy processing in them.
```
Or use `NOTE` in the two occurrences down below, so the casing is consistent.

File: observability/opentelemetry/v2/README.md [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+
+
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md:97]:
```suggestion
```

File: observability/opentelemetry/v2/README.md [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+## Extra types
+
+This package also contains extra types and helper functions that are useful in case you need to access/set the `tracecontext` in a more "low-level" way.
+
+They allow to inject(write) and extract(read) `tracecontext` from the event. This is particularly useful when dealing with code that has no notion of a "request" nor a context. For example, long-running background processes pooling from a queue. 
+
+>Note: To learn more about the propagation, take a look at the [Propagators API SPEC](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.6.1/specification/context/api-propagators.md).
+
+
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
```

File: observability/opentelemetry/v2/README.md [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+## Manually extracting/injecting tracecontext from the event
+
+When working with distributed systems, it can be difficult to achieve proper context propagation. For example, a long-running process listening to a topic does not have a "context" concept like a HTTP server receiving requests does.
+
+>Note: The OpenTelemetry community is always creating auto-instrumentation libraries for popular SDKs and frameworks, so there could be one for the queue client or other framework you are using. It's worth checking out what is available in the [go-contrib GitHub repository](https://github.com/open-telemetry/opentelemetry-go-contrib/tree/main/instrumentation)
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]: Would it also make sense to link to [the OpenTelemetry Registry](https://opentelemetry.io/registry/) here? E.g.: the Queue system has an Auto-instrumentation available out of the box, or would that not work with cloudevents?

File: observability/opentelemetry/v2/README.md [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+	prop.Inject(ctx, carrier)
+
+    // Here then we have the "raw" access to the tracecontext data
+    // https://www.w3.org/TR/trace-context/
+
+    // e.g. 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01
+	me.TraceParent = carrier.Extension.TraceParent
+
+    // e.g. congo=t61rcWkgMzE
+	me.TraceState = carrier.Extension.TraceState
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
type MyEvent struct {
	TraceParent string `json:"traceparent,omitempty"`
	TraceState  string `json:"tracestate,omitempty"`
}

func injectAndReadTraceParentAndState(ctx context.Context, e cloudevents.Event) {

	me := MyEvent{}

	// the propagator from OpenTelemetry
	prop := propagation.TraceContext{}

	carrier := otelObs.NewCloudEventCarrier()

	// Injects (writes) the tracecontext into the NewCloudEventCarrier
	// Doing so, will set the DistributedTracingExtension fields
	prop.Inject(ctx, carrier)

	// Here then we have the "raw" access to the tracecontext data
	// https://www.w3.org/TR/trace-context/

	// e.g. 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01
	me.TraceParent = carrier.Extension.TraceParent

	// e.g. congo=t61rcWkgMzE
	me.TraceState = carrier.Extension.TraceState
}
```

Consistent tab/spaces formatting

File: observability/opentelemetry/v2/client/cloudevents_carrier.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+func NewCloudEventCarrier() CloudEventCarrier {
+	return CloudEventCarrier{Extension: &extensions.DistributedTracingExtension{}}
+}
+
+// NewCloudEventCarrierWithEvent creates a new CloudEventCarrier with a distributed tracing extension
+// populated with the trace data from the event.
+func NewCloudEventCarrierWithEvent(event cloudevents.Event) CloudEventCarrier {
+	var te, ok = extensions.GetDistributedTracingExtension(event)
+	if !ok {
+		return CloudEventCarrier{Extension: &extensions.DistributedTracingExtension{}}
```
Comment (pirgeo) [inline observability/opentelemetry/v2/client/cloudevents_carrier.go:34]: Should there be a warning that the Extension is empty?

File: observability/opentelemetry/v2/http/http.go [hunk 24]
```diff
@@ -0,0 +1,31 @@
...
+// NewObservedHTTP creates an HTTP protocol with OTel trace propagating middleware.
+func NewObservedHTTP(opts ...cehttp.Option) (*cehttp.Protocol, error) {
+	// appends the OpenTelemetry Http transport + Middleware wrapper
+	// to properly trace outgoing and incoming requests from the client using this protocol
+	return cehttp.New(append(
+		[]cehttp.Option{
+			cehttp.WithRoundTripper(otelhttp.NewTransport(http.DefaultTransport)),
+			cehttp.WithMiddleware(func(next http.Handler) http.Handler {
+				return otelhttp.NewHandler(next, "", otelhttp.WithSpanNameFormatter(func(_ string, r *http.Request) string {
+					return "cloudevents.http.receiver." + r.URL.Path
```
Comment (pirgeo) [inline observability/opentelemetry/v2/http/http.go]: Would this create different spans for requests to say `/api/people/1` and `/api/people/2`, etc.? If this is the case, this line might lead to high dimensionality (<https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span>)

File: samples/http/otel-sender-receiver/README.md [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+> If this is your first time hearing about OpenTelemetry, take a look at the official documentation: [What is OpenTelemetry](https://opentelemetry.io/docs/concepts/what-is-opentelemetry/).
+
+The client app sends an event using cloudevents to a server. The server listens for events also using cloudevents. For each event it receives, it makes then an external HTTP request to `cloudevents.io`, just to illustrate the context propagation working. 
+
+Here's a diagram to examplify the communication between the components:
+
+![client > server > external service](./request-flow.png "Sample app request flow")
+
+
```
Comment (pirgeo) [inline samples/http/otel-sender-receiver/README.md]:
```suggestion
```

File: samples/http/otel-sender-receiver/README.md [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+Both apps are configured to send the spans to a locally running [Jaeger](https://www.jaegertracing.io/). See instructions below on how to run everything.
```
Comment (pirgeo) [inline samples/http/otel-sender-receiver/README.md]:
```suggestion
Both apps are configured to send the spans to a locally running [Jaeger](https://www.jaegertracing.io/) instance. See instructions below on how to run everything.
```

File: samples/http/otel-sender-receiver/README.md [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+
+```shell
+$ go run server/server.go
+```
+
+```shell
+$ go run client/client.go
+```
+
+4. Open the Jaeger UI [http://localhost:16686](http://localhost:16686)
```
Comment (pirgeo) [inline samples/http/otel-sender-receiver/README.md]:
```suggestion
4. Open the Jaeger UI (by default Jaeger exposes the UI at [http://localhost:16686](http://localhost:16686))
```

File: samples/http/otel-sender-receiver/README.md [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+
+5. On the `Service` dropdown select `cloudevents-client`. Next, click on `Find Traces`. 
+
+![Selecting the service on Jaeger](./jaeger-find-traces.png "Finding our traces in Jaeger")
+
+
```
Comment (pirgeo) [inline samples/http/otel-sender-receiver/README.md]:
```suggestion
```

Comment (pirgeo) [inline samples/http/otel-sender-receiver/README.md]: Also, the images have a lot of white at the bottom, which makes it look like there is a huge gap below this image.

File: samples/http/otel-sender-receiver/README.md [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+6. On the right side you should see a trace with name `cloudevents-client: cloudevents.client.example.type send`. Click there to see the full trace:
+
+
```
Comment (pirgeo) [inline samples/http/otel-sender-receiver/README.md]:
```suggestion
```

File: observability/opentelemetry/v2/README.md (hunk 17 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/README.md]: No, you are correct 😁

File: observability/opentelemetry/v2/README.md (hunk 21 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/README.md]: AFAIK, the way cloudevents works (at least this one for go) is the `client.go` they offer has a transport that can be many things. HTTP, amqp, nats, kafka, etc. Since there's many protocols, I believe the idea with this Observability service is to always have spans being created when using the client, since it's the same for all protocols, serving as as central point.

I guess listing the Registry also could work. Since this is for Go, I thought of just putting the contrib link but maybe something else goes into the "core" one so it gets missed.

I added auto-instrumentation for HTTP, so incoming/outgoing requests have propagation working. If there are libraries for the other protocols, those could also be added later on for sure.

File: observability/opentelemetry/v2/http/http.go (hunk 24 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/http/http.go]: Yes I also thought about it. They have this as well in their OpenCensus impl: https://github.com/cloudevents/sdk-go/blob/main/observability/opencensus/v2/http/http.go#L25.  I *think* it will not be a path like you mentioned, but just the base path of the listening client (in my tests it was always `/`), which makes it just as bad anyway.


What you think would be a good value? The outgoing events from, the client have a span name by default of `cloudevents.client.<eventtype> <op name>`, e.g. `cloudevents.client.my.event send` which should offer a good low cardinality. 

But in this case it's just a middleware so I don't have the `event` that easily to take the `type`. Maybe then it can just be `cloudevents.http.receiver`?

File: observability/opentelemetry/v2/README.md (hunk 22 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/README.md]: Fixed.

File: observability/opentelemetry/v2/README.md (hunk 21 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/README.md]: I rephrased it a bit stating the registry and as support increases, new auto-instrumentation libraries can be added.

File: observability/opentelemetry/v2/client/cloudevents_carrier.go (hunk 23 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/client/cloudevents_carrier.go:34]: Good point. I added a `warn` log, following how they do it elsewhere.

File: observability/opentelemetry/v2/README.md [hunk 30]
```diff
@@ -0,0 +1,183 @@
+# OpenTelemetry instrumentation for CloudEvents
+
+This package contains the components necessary to instrument CloudEvents clients with OpenTelemetry. The main component is the `OTelObservabilityService` which implements the `ObservabilityService` interface from CloudEvents.
+
+## Getting a fully instrumented CloudEvents HTTP client
```
Comment (pirgeo) [inline observability/opentelemetry/v2/README.md]:
```suggestion
## Instrumented CloudEvents HTTP client
```

**Review by pirgeo — APPROVED** [review]

File: observability/opentelemetry/v2/http/http.go (hunk 24 above)
Comment (joaopgrassi) [reply observability/opentelemetry/v2/http/http.go]: Went with `cloudevents.http.receiver`.
